src/Data/Form.purs

```diff
@@ -55,8 +55,17 @@ type StringRows f r =
   | r
   )
 
+type DropdownRows f r =
+  ( index :: Int
+  , options :: Array { name :: String, value :: f }
```

Comment:
`options` will actually need to be an `Expr`. Which means we'll probably need to add `Array` as a supported `ExprType`.
 
The motivation is so that we can have different options for a dropdown given different form states. E.g.:

```purescript
options:
  (if_ (equal_ (lookup_ "foo") (string_ "B"))
    (array_
      [ pair_ "Foo Bar" (string_ "foo_bar")
      , pair_ "Foo Baz" (string_ "foo_baz")
      ]
    )
    (array_
      [ pair_ "Foo Fiz" (string_ "foo_fiz")
      , pair_ "Foo Wiz" (string_ "foo_wiz")
      ]
    )
  )
```

I dunno that `Pair` is the best constructor for a key/value pair but that was the best I could come up with.

Also, we can get away with nested if/else chains here, but it might be more ergonomic to extend our algebra with `case`s, so we can extend the above as:

```purescript
options:
  (case_ (lookup_ "foo")
    [ (string_ "B") (array_ ...)
    , (string_ "C") (array_ ...)
    ]
    (array_ ...) -- otherwise/failure case
  )
```

Reply:
If you'd like to get a better understanding of the need for dynamic options, try loading up a new campaign form in Wildcat, then changing the `Objective`. You'll notice the available options (as well as the default values) for `Optimization Goal` and `Billing Event` will change (and for some values certain fields like `Application` might show or hide).

https://citizennet.com/manager/#/groups/facebook/159675/campaigns/new/details

Reply:
> Which means we'll probably need to add `Array` as a supported `ExprType`.

If we make arrays `ExprType`s, it doesn't seem like they would be able to have `Expr`s for values. Is that right?

> it might be more ergonomic to extend our algebra with `case`s

Can this be sugar outside the core of the expressions? It seems like we can convert cases to ifs:

```PureScript
case_ :: Expr -> NonEmptyArray (Tuple Expr Expr) -> Expr -> Expr
case_ cond cases def = if_ (equal_ cond ?firstCondition) ?firstResult ?iterateTheRestOfTheCases
```

Reply:
Oh, I guess it could be:
```PureScript
data ExprType
  = Boolean Boolean
  | Int Int
  | String String
  | Array (Array Expr)
```
Nevermind.

Reply:
I'm not certain we need them to be expressions...we might be better off with
```purescript
data ExprType
  = Boolean Boolean
  | Int Int
  | String String
  | Pair Key ExprType
  | Array (Array ExprType)
```

Reply:
And if it turns out we do need to allow for arrays of expressions, we can add support for that once the use-case becomes clear

Reply:
I want to pop up before going down this path. Are we going to implement this for MVP, or are we planning to come back to it afterward? If we're implementing it, we're going to have to also implement `valueRepresentation` (we'll have to traverse the `Expr` to get the value out of it). It doesn't seem _hard_ per say. But it is a non-trivial amount of work.

Reply:
Hmm, why does this force us to implement `valueRepresentation` (side note, need to find a terser name) now? Can't we just render the `name`?

Reply:
I just thought of something. Having `Invalid a` as part of our `InputSource a` coproduct allows us to recover an invalid value if it becomes valid again.

So using my above food example, say `"meal"` is set to `"Breakfast"` and from the current dropdown `"Pancakes"` is selected, so we have `value: UserInput (String "Pancakes")`. The user changes `"meal"` to `"Lunch"` and now we have `value: Invalid (String "Pancakes")` (but the `default` is shown in the dropdown if one exists). Then the user switches back to `"Breakfast"`, suddenly our invalid value is valid again, so we should recover it and end back up with `value: UserInput (String "Pancakes")`.

Reply:
> Hmm, why does this force us to implement `valueRepresentation` (side note, need to find a terser name) now? Can't we just render the `name`?

I don't think it's a hard requirement,  but whatever work we'd do in PS to grab the `name` out of the `Expr` would be equivalent to us implementing `valueRepresentation` and using it to do the work. Thinking about it a bit more, how are we going to do `ExprType -> Array ExprType` for the dropdown component's [`items`][items]?

If we go back to the impetus for this change, we want options to be `Expr` so we can change the options on the fly. Could we instead have different dropdowns and let the `visibility` decide when it's on the form?

Let's say we have `options :: Array { name :: String, value :: Expr }` or `options :: Array Expr`. and the `Objective` changes, it has to be cased on in the `visibility` for every field,  but they do no further casing on the Objective for `options`, `placeholder`, or `valueRepresentation`. Someone implementing a form has multiple dropdowns for each `Objective`.

For example:

```PureScript
foo :: Field Expr
foo =
  { name: string_ "foo"
  , visibility: equal_ (string_ "objective-foo") (lookup_ "objective" (boolean_ false))
  , description: string_ "foo"
  , key: "foo"
  , input: Dropdown
    { default: Nothing
    , index: 0
    , options:
      [ { name: "foo-1", value: string_ "foo-1" }
      , { name: "foo-2", value: string_ "foo-2" }
      , { name: "foo-3", value: string_ "foo-3" }
      ]
    , placeholder: string_ ""
    , required: boolean_ true
    , value: NotSet
    , valueRepresentation: string_ ""
    }
  }

bar :: Field Expr
bar =
  { visibility: equal_ (string_ "objective-bar") (lookup_ "objective" (boolean_ false))
  , description: string_ "bar"
  , key: "bar"
  , input: Dropdown
    { default: Nothing
    , index: 0
    , options:
      [ { name: "bar-1", value: string_ "bar-1" }
      , { name: "bar-2", value: string_ "bar-2" }
      , { name: "bar-3", value: string_ "bar-3" }
      ]
    , placeholder: string_ ""
    , required: boolean_ true
    , value: NotSet
    , valueRepresentation: string_ ""
    }
  }

baz :: Field Expr
baz =
  { visibility: equal_ (string_ "objective-baz") (lookup_ "objective" (boolean_ false))
  , description: string_ "baz"
  , key: "baz"
  , input: Dropdown
    { default: Nothing
    , index: 0
    , options:
      [ { name: "baz-1", value: string_ "baz-1" }
      , { name: "baz-2", value: string_ "baz-2" }
      , { name: "baz-3", value: string_ "baz-3" }
      ]
    , placeholder: string_ ""
    , required: boolean_ true
    , value: NotSet
    , valueRepresentation: string_ ""
    }
  }
```

Let's say we have `options :: Expr`. If the `Objective` changes, it has to be cased on in the `visibility` for every field (while handling multiple cases), and it has to be cased on within the dropdown (`options`, `placeholder`, and `valueRepresentation`). Someone implementing the form has one large dropdown expression with many cases.

For example:
```PureScript
fooBarBaz :: Field Expr
fooBarBaz =
  { name:
    if_
      (equal_ (string_ "objective-foo") (lookup_ "objective" (boolean_ false)))
      (string_ "foo")
      (if_
        (equal_ (string_ "objective-bar") (lookup_ "objective" (boolean_ false)))
        (string_ "bar")
        (string_ "baz"))
  , visibility:
    if_
      (equal_ (string_ "objective-foo") (lookup_ "objective" (boolean_ false)))
      (boolean_ true)
      (if_
        (equal_ (string_ "objective-bar") (lookup_ "objective" (boolean_ false)))
        (boolean_ true)
        (equal_ (string_ "objective-baz") (lookup_ "objective" (boolean_ false))))
  , description:
    if_
      (equal_ (string_ "objective-foo") (lookup_ "objective" (boolean_ false)))
      (string_ "foo")
      (if_
        (equal_ (string_ "objective-bar") (lookup_ "objective" (boolean_ false)))
        (string_ "bar")
        (string_ "baz"))
  , key: "fooBarBaz"
  , input: Dropdown
    { default: Nothing
    , index: 0
    , options:
      if_
        (equal_ (string_ "objective-foo") (lookup_ "objective" (boolean_ false)))
        (array_
         [ pair_ "foo-1" (string_ "foo-1")
         , pair_ "foo-2" (string_ "foo-2")
         , pair_ "foo-3" (string_ "foo-3")
         ])
        (if_
          (equal_ (string_ "objective-bar") (lookup_ "objective" (boolean_ false)))
          (array_
            [ pair_ "bar-1" (string_ "bar-1")
            , pair_ "bar-2" (string_ "bar-2")
            , pair_ "bar-3" (string_ "bar-3")
            ])
          (array_
            [ pair_ "baz-1" (string_ "baz-1")
            , pair_ "baz-2" (string_ "baz-2")
            , pair_ "baz-3" (string_ "baz-3")
            ]))
    , placeholder: string_ ""
    , required: boolean_ true
    , value: NotSet
    , valueRepresentation: string_ ""
    }
  }
```

The former seems easier for us and for our users. Did I oversimplify things? If that does work, it also seems like it simplifies typeahead a bit as well, since it would be an `Array Expr` as well and would fall under the same rules.

[items]: https://github.com/citizennet/purescript-ocelot/blob/master/src/Components/Dropdown/Component.purs#L30

Reply:
For audit, we discussed live and I did over simplify it. There are some cases we have where it makes sense to have one dropdown with repeated casing. We're going on with the `options :: Expr`.